File: ios-lister/CloudStoreLister/Logic/ListSortLogic.swift
import Foundation

enum ListSortField: String, CaseIterable, Identifiable {
    case name = "Name"
    case productType = "Type"
    case productId = "Product ID"
    case barcode = "Barcode"
    case price = "Price"
    case stock = "Stock"
    case pullCount = "Pull Count"

    var id: String { rawValue }
}

enum ListSortLogic {
    static func sorted(
        _ items: [InventoryListItem],
        by field: ListSortField,
        ascending: Bool
    ) -> [InventoryListItem] {
        items.sorted { lhs, rhs in
            let ordered = compare(lhs, rhs, by: field)
            if ordered == .orderedSame {
                return lhs.productId < rhs.productId
            }
            return ascending ? ordered == .orderedAscending : ordered == .orderedDescending
        }
    }

    static func sortList(
        id: UUID,
        by field: ListSortField,
        ascending: Bool,
        in lists: [InventoryNamedList]
    ) -> [InventoryNamedList]? {
        guard let index = lists.firstIndex(where: { $0.id == id }),
              !lists[index].items.isEmpty else { return nil }
        var updated = lists
        updated[index].items = sorted(updated[index].items, by: field, ascending: ascending)
        return updated
    }

    private static func compare(
        _ lhs: InventoryListItem,
        _ rhs: InventoryListItem,
        by field: ListSortField
    ) -> ComparisonResult {
        switch field {
        case .name:
            return lhs.name.localizedCaseInsensitiveCompare(rhs.name)
        case .productType:
            return (lhs.productType ?? "").localizedCaseInsensitiveCompare(rhs.productType ?? "")
        case .productId:
            return numericCompare(lhs.productId, rhs.productId)
        case .barcode:
            return (lhs.barcode ?? "").localizedCaseInsensitiveCompare(rhs.barcode ?? "")
        case .price:
            return lhs.priceLabel.localizedCaseInsensitiveCompare(rhs.priceLabel)
        case .stock:
            return lhs.stockLabel.localizedCaseInsensitiveCompare(rhs.stockLabel)
        case .pullCount:
            return numericCompare(lhs.pullCount, rhs.pullCount)
        }
    }

    private static func numericCompare<T: Comparable>(_ lhs: T, _ rhs: T) -> ComparisonResult {
        if lhs < rhs { return .orderedAscending }
        if lhs > rhs { return .orderedDescending }
        return .orderedSame
    }
}

File: android-lister/app/src/main/java/com/cloudstore/lister/domain/ListSortLogic.kt
package com.cloudstore.lister.domain

import com.cloudstore.lister.data.InventoryListItem
import com.cloudstore.lister.data.InventoryNamedList

enum class ListSortField(val label: String) {
    Name("Name"),
    ProductType("Type"),
    ProductId("Product ID"),
    Barcode("Barcode"),
    Price("Price"),
    Stock("Stock"),
    PullCount("Pull Count"),
}

object ListSortLogic {
    fun sortList(
        id: String,
        field: ListSortField,
        ascending: Boolean,
        lists: List<InventoryNamedList>,
    ): List<InventoryNamedList>? {
        val index = lists.indexOfFirst { it.id == id }
        if (index < 0 || lists[index].items.isEmpty()) return null
        return lists.toMutableList().also {
            it[index] = it[index].copy(items = sorted(it[index].items, field, ascending))
        }
    }

    fun sorted(items: List<InventoryListItem>, field: ListSortField, ascending: Boolean): List<InventoryListItem> {
        return items.sortedWith { lhs, rhs ->
            val cmp = compare(lhs, rhs, field)
            when {
                cmp == 0 -> lhs.productId.compareTo(rhs.productId)
                ascending -> cmp
                else -> -cmp
            }
        }
    }

    private fun compare(lhs: InventoryListItem, rhs: InventoryListItem, field: ListSortField): Int {
        return when (field) {
            ListSortField.Name -> lhs.name.compareTo(rhs.name, ignoreCase = true)
            ListSortField.ProductType -> (lhs.productType ?: "").compareTo(rhs.productType ?: "", ignoreCase = true)
            ListSortField.ProductId -> lhs.productId.compareTo(rhs.productId)
            ListSortField.Barcode -> (lhs.barcode ?: "").compareTo(rhs.barcode ?: "", ignoreCase = true)
            ListSortField.Price -> lhs.priceLabel.compareTo(rhs.priceLabel, ignoreCase = true)
            ListSortField.Stock -> lhs.stockLabel.compareTo(rhs.stockLabel, ignoreCase = true)
            ListSortField.PullCount -> lhs.pullCount.compareTo(rhs.pullCount)
        }
    }
}
